### PR TITLE
Project Service - Project Staff - CRUD

### PR DESCRIPTION
**Pull request by jitendrac-egov**
Added

Project Staff Creation, Update, Delete and search

Health Common Services added the User Service

### CODE REVIEW

File: health-services/project/src/main/java/org/egov/project/repository/ProjectRepository.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+    private final NamedParameterJdbcTemplate namedParameterJdbcTemplate;
+
+    @Autowired
+    public ProjectRepository(NamedParameterJdbcTemplate namedParameterJdbcTemplate) {
+        this.namedParameterJdbcTemplate = namedParameterJdbcTemplate;
+    }
+
+    public List<String> validateProjectIds(List<String> projectIds) {
+        String query = String.format(
```
Comment (jayantp-egov) [inline health-services/project/src/main/java/org/egov/project/repository/ProjectRepository.java:26]: Use cache here

File: health-services/project/src/main/java/org/egov/project/repository/ProjectRepository.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+                "SELECT id FROM project WHERE id IN (:projectIds) AND isDeleted = false fetch first %s rows only",
+                projectIds.size());
+        Map<String, Object> paramMap = new HashMap<>();
+        paramMap.put("projectIds", projectIds);
+        try {
+            return namedParameterJdbcTemplate.queryForList(query, paramMap, String.class);
+        } catch (EmptyResultDataAccessException e) {
```
Comment (jayantp-egov) [inline health-services/project/src/main/java/org/egov/project/repository/ProjectRepository.java:33]: Remove this exception, refer dev branch product and product variant code

File: health-services/project/src/main/java/org/egov/project/repository/ProjectStaffRepository.java [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+@Slf4j
+@Repository
+public class ProjectStaffRepository {
+
+    private static final String HASH_KEY = "project-staff";
```
Comment (jayantp-egov) [inline health-services/project/src/main/java/org/egov/project/repository/ProjectStaffRepository.java:28]: Add new line after every declaration

File: health-services/project/src/main/java/org/egov/project/repository/ProjectStaffRepository.java [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+    }
+
+    public List<ProjectStaff> find(ProjectStaffSearch projectStaffSearch,
+                                   Integer limit,
+                                   Integer offset,
+                                   String tenantId,
+                                   Long lastChangedSince,
+                                   Boolean includeDeleted) throws QueryBuilderException {
+        String query = selectQueryBuilder.build(projectStaffSearch);
+        query += " and tenantId=:tenantId ";
```
Comment (jayantp-egov) [inline health-services/project/src/main/java/org/egov/project/repository/ProjectStaffRepository.java:119]: Put proper spaces, refer to latest code from dev branch

File: health-services/project/src/main/java/org/egov/project/repository/ServiceRequestRepository.java [hunk 5]
```diff
@@ -6,6 +6,7 @@
 import lombok.extern.slf4j.Slf4j;
 import org.egov.tracer.model.ServiceCallException;
 import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
```
Comment (jayantp-egov) [inline health-services/project/src/main/java/org/egov/project/repository/ServiceRequestRepository.java:9]: Remove this file. Not needed.

File: health-services/project/src/main/java/org/egov/project/repository/rowmapper/ProjectStaffRowMapper.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    public ProjectStaff mapRow(ResultSet resultSet, int i) throws SQLException {
+        try {
+            return ProjectStaff.builder()
+                    .id(resultSet.getString("id"))
+                    .tenantId(resultSet.getString("tenantid"))
+                    .projectId(resultSet.getString("projectId"))
+                    .userId(resultSet.getString("staffId"))
+                    .startDate(resultSet.getLong("startDate"))
+                    .endDate(resultSet.getLong("endDate"))
+                    .additionalFields(objectMapper.readValue(resultSet.getString("additionalDetails"), AdditionalFields.class))
```
Comment (jayantp-egov) [inline health-services/project/src/main/java/org/egov/project/repository/rowmapper/ProjectStaffRowMapper.java:27]: You will have to write some conversion logic(null check) for additionalDetails. Refer to dev branch.

File: health-services/project/src/test/java/org/egov/project/repository/ProjectRepositoryTest.java [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+    void shouldReturnListOfValidProjectIdsGivenAListOfProjectIds() {
+        List<String> projectIds = Arrays.asList("project1", "project2", "project3");
+        Map<String, Object> params = new HashMap<>();
+        params.put("projectIds", projectIds);
+
+        when(
+                namedParameterJdbcTemplate.queryForList(
+                eq(
+                        String.format(
+                            "SELECT id FROM project WHERE id IN (:projectIds) AND isDeleted = false fetch first %s rows only",
```
Comment (jayantp-egov) [inline health-services/project/src/test/java/org/egov/project/repository/ProjectRepositoryTest.java:46]: Correct formatting

**Review by jayantp-egov — CHANGES_REQUESTED** [review]

File: health-services/project/src/main/java/org/egov/project/repository/ProjectStaffRepository.java (hunk 3 above)
Comment (jitendrac-egov) [reply health-services/project/src/main/java/org/egov/project/repository/ProjectStaffRepository.java:28]: Added

File: health-services/libraries/health-services-common/src/main/java/org/egov/common/service/UserService.java [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+            UserDetailResponse response =  restRepo.fetchResult(
+                    new StringBuilder(host + searchUrl),
+                    userSearchRequest,
+                    UserDetailResponse.class
+            );
+
+            return response.getUser();
+
+        } catch (Exception e) {
+            log.error("Exception while searching users : " + e.getMessage());
```
Comment (jayantp-egov) [inline health-services/libraries/health-services-common/src/main/java/org/egov/common/service/UserService.java:48]: log.error("message", e)

**Review by jayantp-egov — APPROVED** [review]

**Pull request merged**